GCEWindowsAgent/addresses.go

```diff
@@ -62,6 +65,152 @@ func (a *addressMgr) parseWSFCEnable() bool {
+}
+
+func getRoutes(ifname string) ([]string, error) {
+	args := fmt.Sprintf("route list table local type local scope host dev %s proto %d", ifname, protoID)
+	out, err := exec.Command("ip", strings.Split(args, " ")...).Output()
+	if err != nil {
+		if ee, ok := err.(*exec.ExitError); ok {
+			return nil, fmt.Errorf("Error getting routes: %s", ee.Stderr)
+		}
+		return nil, err
+	}
+	var res []string
+	for _, line := range strings.Split(string(out), "\n") {
+		line = strings.TrimPrefix(line, "local ")
+		line = strings.TrimSpace(line)
+		if line != "" {
+			res = append(res, line)
+		}
+	}
+	return res, nil
+}
+
+func addRoute(ip, ifname string) error {
+	if strings.Index(ip, "/") == -1 {
+		ip = ip + "/32"
+	}
+	args := fmt.Sprintf("route add to local %s scope host dev %s proto %d", ip, ifname, protoID)
+	_, err := exec.Command("ip", strings.Split(args, " ")...).Output()
+	if ee, ok := err.(*exec.ExitError); ok {
+		return fmt.Errorf("Error adding route: %s", ee.Stderr)
+	}
+	return err
+}
+
+func removeRoute(ip, ifname string) error {
+	if strings.Index(ip, "/") == -1 {
+		ip = ip + "/32"
+	}
+	args := fmt.Sprintf("route delete to local %s scope host dev %s proto %d", ip, ifname, protoID)
+	_, err := exec.Command("ip", strings.Split(args, " ")...).Output()
+	if ee, ok := err.(*exec.ExitError); ok {
+		return fmt.Errorf("Error removing route: %s", ee.Stderr)
+	}
+	return err
+}
+
+// Filter out forwarded ips based on WSFC (Windows Failover Cluster Settings).
+// If only EnableWSFC is set, all ips in the ForwardedIps and TargetInstanceIps will be ignored.
+// If WSFCAddresses is set (with or without EnableWSFC), only ips in the list will be filtered out.
+func (a *addressMgr) applyWSFCFilter() {
+	wsfcAddresses := a.parseWSFCAddresses()
+
+	var wsfcAddrs []string
+	for _, wsfcAddr := range strings.Split(wsfcAddresses, ",") {
+		if wsfcAddr == "" {
+			continue
+		}
+
+		if net.ParseIP(wsfcAddr) == nil {
+			logger.Errorf("ip address for wsfc is not in valid form %s", wsfcAddr)
+			continue
+		}
+
+		wsfcAddrs = append(wsfcAddrs, wsfcAddr)
+	}
+
+	if len(wsfcAddrs) != 0 {
+		interfaces := newMetadata.Instance.NetworkInterfaces
+		for idx := range interfaces {
+			var filteredForwardedIps []string
+			for _, ip := range interfaces[idx].ForwardedIps {
+				if !containsString(ip, wsfcAddrs) {
+					filteredForwardedIps = append(filteredForwardedIps, ip)
+				}
+			}
+			interfaces[idx].ForwardedIps = filteredForwardedIps
+
+			var filteredTargetInstanceIps []string
+			for _, ip := range interfaces[idx].TargetInstanceIps {
+				if !containsString(ip, wsfcAddrs) {
+					filteredTargetInstanceIps = append(filteredTargetInstanceIps, ip)
+				}
+			}
+			interfaces[idx].TargetInstanceIps = filteredTargetInstanceIps
+		}
+	} else {
+		wsfcEnable := a.parseWSFCEnable()
+		if wsfcEnable {
+			for idx := range newMetadata.Instance.NetworkInterfaces {
+				newMetadata.Instance.NetworkInterfaces[idx].ForwardedIps = nil
```

Comment:
this logic means that global var newMetadata is updated in place, meaning all future diffs will trigger updates. essentially, if WSFC is enabled, we will run this manager once a minute regardless of changes to metadata.

probably this function should take and return a metadata object so that the global isn't changed unless caller decides it should be. but i didn't want to change too much of the function of this code if i didn't have to in this PR.